### DISCUSSION TITLE
espnow receaver how to differentiate data

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#839] [Q&A] [answered]
I  am sending data of different encooder  to  espnow . Mostly sbc , mp3 and wave . 

But at receiver end how could i DFFERENTATE which encode type data it is so to write it correct decoder

**Comment by pschatzmann** [comment, marked as answer]
You can use my [BinaryContainer](https://www.pschatzmann.ch/home/2023/04/28/arduino-audio-tools-codecs-and-containers/) and send the format as metadata

**Reply by aqildad-create** [reply]
you don't have any example for writeMeta .  If you provide some it will highly appreciated .

**Reply by aqildad-create** [reply]
and if i use encoder with binary container and copystream to espnow then does it will effect the speed or data or timing of espnow  ?? because as per my understanding binary container will add some data with encoded data and this may add additional data load on espnow .

**Reply by pschatzmann** [reply]
The overhead is just minimal and I would expect that it can be ignored...

**Reply by aqildad-create** [reply]
I am getting error n container file 
![image](https://github.com/pschatzmann/arduino-audio-tools/assets/65456998/0252d3c5-9f0a-47dc-8dc0-9043b3322caa)

**Reply by pschatzmann** [reply]
I just changed the enums to enum classes to avoid any conflicts

**Comment by pschatzmann** [comment]
Alternatively you can also just analyse the file format! StreamCopy contains some basic logic which should recognise mp3, wav and aac

**Reply by aqildad-create** [reply]
I wanted to analyze file format via streamCopy but still i am confuse and were can i get documentation of these basic logics as you mention  .

**Comment by pschatzmann** [comment]
No, just look at the source code: You can implement a similar logic directly in your sketch yourself:
https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioTools/StreamCopy.h

Just have a look at notifyMime

**Reply by aqildad-create** [reply]
notifyMime can only analyze  4 audio formats (audio/basic ,audio/aac  , audio/mpeg  , audio/vnd.wave )  . rights ? and mp3 is not present may be because aac and mp3 is same correct ? ? 

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/65456998/4f9a5c8f-1212-494c-a21e-2ceffb4a7d64)


Further what about sbc formate  ? It can achieve like this If none of audio formate is present then  It must be SBC correct ? because currently  have wav  mp3 and sbc formats only  .   am I correct ?  

Further for implementation point of view   first i will  read data from now stream like this 
ESPNowStream now;
.
now.readBytes( . . 


then pass to notifymine for first time only   correct ?? 

and at last at receiver end of espnow how could  I know that audio has completed or new format data  is coming  so that i will check with notifymine again ?

**Comment by pschatzmann** [comment]
The mime type for mp3 is audio/mpeg

**Comment by aqildad-create** [comment]
And how to know again that if new formate data is coming because I will
check only fist time espnow data receive .. How will I know again data
formate has changed ?

**Reply by aqildad-create** [reply]
below is the implementaton  I try  and continuous check in the loop because  i don't know when the transmitter will change the data format . 
for testing i just serial out the format . 

Is this correct way of finding the format at receiver end of espnow  ? 
.
.
.
I2SStream out;
MultiOutput MyOut;
RingBufferStream test;
EncodedAudioStream decoder(&out, new SBCDecoder( ));
StreamCopy copier(MyOut, now);
. 
MyOut.add(decoder);
MyOut.add(test);

.
loop()
{
copier.copy();
  int copied = test.readBytes((uint8_t *)myBufferbuffer , 10);
   test.flush();
   notifyMime(myBufferbuffer ,copied );
}

 void notifyMime(void* data, size_t len){
            if (    len>4) {
                const uint8_t *start = (const uint8_t *) data;
                actual_mime = "audio/basic";
                if (start[0]==0xFF && start[1]==0xF1){
                    actual_mime = "audio/aac";
                } else if (memcmp(start,"ID3",3) || start[0]==0xFF || start[0]==0xFE ){
                    actual_mime = "audio/mpeg";
                } else if (memcmp(start,"RIFF",4)){
                    actual_mime = "audio/vnd.wave";
                }
               
            }
            Serial.println(actual_mime);
        }
